Std/Tactic/Ext.lean

```diff
@@ -6,9 +6,30 @@ Authors: Gabriel Ebner, Mario Carneiro
 import Std.Tactic.Basic
 import Std.Tactic.RCases
 import Std.Tactic.Ext.Attr
+import Std.Tactic.TryThis
 
 namespace Std.Tactic.Ext
-open Lean Meta Elab Tactic
+open Lean Meta Elab Tactic TryThis
+
+/-- Enables tracing in the `ext` tactic, displaying used extensionality lemmas.
+This option is used by `ext?` and `ext1?`. -/
+register_option tactic.ext.trace : Bool := {
+  defValue := false
+  descr    := "When enabled, calls to `ext` will print a tactic sequence
+  reproducing the `ext` call."
+}
+
+/-- Forces `ext` to ignore the `DiscTree` and try all `ext`-lemmas.
+This option is used by `ext!?` -/
+register_option tactic.ext.force : Bool := {
+  defValue := false
+  descr    := "When enabled, calls to `ext` will try any `ext`-lemma ignoring the
+  `DistrTree` keys."
+}
+
+/-- Global variable tracking tactics used by `ext?`. This is a collection of
+`apply $lem`, `rintro $pat`, `intros`, and `sorry` statements. -/
+initialize usedExtTactics : IO.Ref (Array (Syntax.Tactic)) ← IO.mkRef #[]
```

Comment:
I'm a bit confuse why these are global settings; can't they just be arguments to `ext`?

Reply:
There are indeed two options here, either use options or pass the settings through all the nested functions. If you think the latter is better, that should not be a hard refactor.

I took inspiration from `simp?` which is implemented this way in core:
https://github.com/leanprover/lean4/blob/f1412ddb4550c22a4302542e4191c54af0cf758e/src/Lean/Elab/Tactic/Simp.lean#L247-L250

Same for `usedExtTactics`: You could also pass it back and forth through these nested function calls, I just thought that would be more complicated as it would change all the signatures of them. And it would mean the trace is even passed around in the 99.99% of times you just use `ext`

Reply:
Maybe @digama0 can comment, but I suspect that `simp?` is not necessarily the model to follow: Mario's design considerations may have been in part dealing with the difficulties of getting changes to core refactored, so rather than replumbing anything he worked around the edges.

It would probably fit better for the Mathlib/Std style to avoid using options and `Ref`s to pass state for tactics.

The trace could be returned inside an `Option`, which is `none` when not needed?

Reply:
Using `IO.Ref` is not the same as using a global ref though. I would avoid global refs if possible, because they have issues in multithreaded situations.